GUIDELINES.md

```diff
@@ -0,0 +1,135 @@
+  * In case the steps class' aim is purely to offer and act on a fake implementation of a class, name it with a `FakeSteps` suffix (e.g. `BackendGatewayFakeSteps` or `SessionStoreFakeSteps`)
+* Start with the test
+  * In the test class you created add reference to the steps class with the `steps` function (e.g. `val sut by steps<LoginSteps>`) and call the value `sut` (system under test)
+  * Fill up the test with `@Test`-annotated test functions
+    * The test function can be named after a _scenario_ (e.g. `New user trying to log in`, `Weather condition is cloudy`)
+    * Only if necessary to distinguish test cases, also add more context like what's roughly expected (e.g. `New user trying to login - welcome to the app message shown`, `Existing user trying to login - welcome back message shown`)
+    * Don't use names like `givenNewUser_whenLoggingIn_thenShowWelcomeToAppMessage` because
+      * The test code in the test functions should already be easy enough to understand preconditions, actions and assumptions
+      * The more text you use the harder to distinguish test cases
+      * The less spaces you use the harder to read
+      * So please just use the test function names to make a meaningful distinctions to properly navigate between and understand test scenarios and cases
+  * For each test case, call functions that would reside in the steps class and prefix these functions with `given`, `when` or `then`
+    * Functions starting with `given` should set up the environment, as these define the assumptions that are made (e.g. `givenUser(email = "test@test.com", password = "supersecure")` indicating that the fake backend should act as if there was that user saved in the database). For the naming choose present tense (e.g. `givenUser` as a short form of "there is a user" or `givenInternetConnected`)
+    * Functions starting with `when` should perform actions, therefore they are named in progressive form (e.g. `whenGoingOffline`, `whenInitializing`, `whenLoggingInWith(email, password)`)
+    * Functions starting with `then` should perform assertions (e.g. `thenIsLoggedIn()`, `thenIsBusy()` or `thenUserIsCreated(user: User)`)
+    * By starting at the test to define which functions are needed in the steps you are more focused and get a faster overview of that's needed instead of getting into details too early
+  * Now the test should already be human-readable for non-tech people (with just the technical implementation in the steps class missing)
+  * Remember that in the test class all data structures and function calls should be domain-specific and therefore be modelled around real-world domain objects and data visible to a user, as far as possible (if the test subject is of total technical nature this doesn't apply of course)
+* Now think about which dependencies (classes or interfaces) will be under test and add new steps classes
+  * You can add a steps classs per class under test (e.g. `AuthManagerSteps`)
```

Comment:
Would you really create steps class per class or more per Sub-System?
Because just thinking, on a higher level (where the tests are written), I wouldn't care about an `AuthManager`, what I care about is e.g. does the user exist on the backend → `FakeBackendSteps`, etc.

Reply:
i think you're right. as we consider steps classes as a way of abstracting technical implementation these should probably be more independent of the _concrete_ implementation and structure. i think it depends on the case. by having steps classes like `LoginViewModelIntegrationSteps` we integrate the whole system anyways already. changes in the structure should not affect the test class, only the steps class, in an optimal case. so according to your input i'd call it `LoginIntegrationSteps` or simply `LoginSteps`?

Reply:
proposal:

There are multiple options how steps classes can be named; some examples:

1. As abstraction of a class (`LoginViewModelSteps`): use this if a steps class solely concentrates on interacting with or mocking/faking a single specific class or interface 
2. As abstraction of an integration of classes (`LoginViewModelIntegrationSteps`): in cases where the class or interface is tested in integration with other classes
3. As abstraction of a feature or subsystem (`LoginSteps`): in cases where

Reply:
Maybe add a naming conventions table at the bottom, as mentioned above already.